### PR TITLE
fix(ecma_parser): 🐛 #5752

### LINKED ISSUE #5752: `delete` used with parentheses always incorrectly leads to TS2703
### Describe the bug

```typescript
let foo: { bar?: number } = { bar: 123 }
delete (foo.bar);
```

The above codes [worked in tsc](https://www.typescriptlang.org/play?#code/DYUwLgBAZg9jBcEDeEBGBDATgfkQOwFcBbVETCAXwgF5k0tEBGAJgGZKAoAExFDBAgAKWDAB0GTAEoA3EA), but [got the following output in swc](https://play.swc.rs/?version=1.2.247&code=H4sIAAAAAAAAA8tJLVFIy8%2B3UqhWSEossrdSyCvNTUotUqhVsIUIWSkYGhkr1HKlpOaklqQqaAAV6wGFNa0BJpypLzoAAAA%3D&config=H4sIAAAAAAAAA0WNSwrDMAxE76J1tt3kDj2ESZXg4B8aBWqM7165OGQnzedNoxMbrY2KE7CMCzWp%2B9JKWgtjE1%2BUFlKYtLsA7vY4OVgtwniZF3IGT3eh6JPf6yBtORZh4LFcOsKd7AaK%2BXMNof3HDGiVmNMJ6g9olj3eM61ycf8BnwbyVboAAAA%3D):

```
  x The operand of a delete operator must be a property reference.
   ,----
 2 | delete (foo.bar);
   :         ^^^^^^^
   `----


Caused by:
    0: failed to parse code
    1: error was recoverable, but proceeding would result in wrong codegen
    2: Syntax Error
```

### Input code

_No response_

### Config

_No response_

### Playground link

https://play.swc.rs/?version=1.2.247&code=H4sIAAAAAAAAA8tJLVFIy8%2B3UqhWSEossrdSyCvNTUotUqhVsIUIWSkYGhkr1HKlpOaklqQqaAAV6wGFNa0BJpypLzoAAAA%3D&config=H4sIAAAAAAAAA0WNSwrDMAxE76J1tt3kDj2ESZXg4B8aBWqM7165OGQnzedNoxMbrY2KE7CMCzWp%2B9JKWgtjE1%2BUFlKYtLsA7vY4OVgtwniZF3IGT3eh6JPf6yBtORZh4LFcOsKd7AaK%2BXMNof3HDGiVmNMJ6g9olj3eM61ycf8BnwbyVboAAAA%3D

### Expected behavior

Should not report any errors.

### Actual behavior

_No response_

### Version

1.2.247

### Additional context

I found that the TS2703 error is thrown by `swc_ecma_parser/src/parser/expr/ops.rs`:

```rust
match &*arg {
    Expr::Member(..) => {}
    Expr::OptChain(OptChainExpr {
        base: OptChainBase::Member(..),
        ..
    }) => {}
    _ => self.emit_err(unwrap_paren(&arg).span(), SyntaxError::TS2703),
}
```

Here the branches does not take `Paren`s into account.

Would love to open a PR for this if it is confirmed as a bug!

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
**Description:**
1. Fixing parser issue about `delete valid parenthesis expr`

**BREAKING CHANGE:**

**Related issue (if exists):**
Closes #5752

### CODE REVIEW

File: crates/swc/tests/tsc-references/deleteChain.1.normal.js [hunk 1]
```diff
@@ -1,37 +1,15 @@
 //// [deleteChain.ts]
```
Comment (IWANABETHATGUY) [inline crates/swc/tests/tsc-references/deleteChain.1.normal.js:1]: https://www.typescriptlang.org/play?ts=4.8.0-beta#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoHATLRAGwENYtREB7AOxlGoEY0NbCAzeWrfUAH1ABvUACM0MBLQDmoAL4BuAlmJZolZgH4AdKKWFV60AAotugJRLlJcpRr1ojAEyt2WLjz6CR44VQlw3LJy8voqahpOOqLaiGGGlKZRurGWeIQ2FFR0DNQAzGg+rpzcvAJ+iAFSwaHKCYx5ujpxdREm+U2pVhlkWfa5ACyFYprDlX74o35YVUHy8yGKrUbUA52I2pPaWPFtpmuizZvmOjvLiavrm6dp1r12OY7UAKzDh8bmY8OTwzOgknNFgtditnp0PrFrtsQRcwYdtBCNlssLcerZsg5GAA2N5TETjEQ-aazGQLYHnbGdADaAHJEDSALpQ2lYRkw9pY6l0xnMmmshmWIA
Typescript still emits some errors, but not `The operand of a delete operator must be a property reference.`
It is related to 
```ts
 function checkDeleteExpressionMustBeOptional(expr: AccessExpression, symbol: Symbol) {
    const type = getTypeOfSymbol(symbol);
    if (strictNullChecks &&
        !(type.flags & (TypeFlags.AnyOrUnknown | TypeFlags.Never)) &&
        !(exactOptionalPropertyTypes ? symbol.flags & SymbolFlags.Optional : getTypeFacts(type) & TypeFacts.IsUndefined)) {
        error(expr, Diagnostics.The_operand_of_a_delete_operator_must_be_optional);
    }
}
```

Comment (IWANABETHATGUY) [reply crates/swc/tests/tsc-references/deleteChain.1.normal.js:1]: It is related to type information, so I think we could ignore them. As Babel did the same thing. https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=PTAEAEGcBcCcEsDG0Bco4FcCmAoHATLRAGwENYtREB7AOxlGoEY0NbCAzeWrfUAH1ABvUACM0MBLQDmoAL4BuAlmJZolZgH4AdKKWFV60AAotugJRLlJcpRr1ojAEyt2WLjz6CR44VQlw3LJy8voqahpOOqLaiGGGlKZRurGWeIQ2FFR0DNQAzGg-rpzcvAJ-iAFSwaHKCYx5ujpxdREm-U2pVhlkWfa5ACyFYprDlX74o35YVUHy8yGKrUbUA52I2pPaWPFtpmuizZvmOjvLiavrm6dp1r12OY7UAKzDh8bmY8OTwzOgknNFgtditnp0PrFrtsQRcwYdtBCNlssLcerZsg5GAA2N5TETjEQ_aazGQLYHnbGdADaAHJEDSALpQ2lYRkw9pY6l0xnMmmshmWIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ctypescript&prettier=false&targets=&version=7.18.13&externalPlugins=&assumptions=%7B%7D

File: crates/swc_ecma_parser/src/parser/expr/ops.rs [hunk 2]
```diff
@@ -294,19 +294,28 @@ impl<I: Tokens> Parser<I> {
             }
 
             if self.input.syntax().typescript() && op == op!("delete") {
-                fn unwrap_paren(e: &Expr) -> &Expr {
-                    match *e {
-                        Expr::Paren(ref p) => unwrap_paren(&p.expr),
-                        _ => e,
+                fn unwrap_paren_iterate(e: &Expr) -> &Expr {
```
Comment (kdy1) [inline crates/swc_ecma_parser/src/parser/expr/ops.rs:297]: I think this is identical as https://rustdoc.swc.rs/swc_ecma_ast/enum.Expr.html#method.unwrap_parens

Comment (kdy1) [reply crates/swc_ecma_parser/src/parser/expr/ops.rs:297]: Maybe we should change it to iterate, instead of recursing.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_ast
 - swc_ecma_parser

**Pull request merged**